biz.aQute.bndlib/src/aQute/bnd/build/WorkspaceLock.java

```diff
@@ -103,6 +104,30 @@ private TimeoutException timeout(Lock lock) {
 		return e;
 	}
 
+	<T, U> T writeReadLocked(final long timeoutInMs, final Callable<U> underWrite,
+		final FunctionWithException<U, T> underRead, final BooleanSupplier canceled) throws Exception {
+		Callable<U> writeLocked = () -> {
+			U writeResult = underWrite.call();
+			// It is safe to let an exception propagate from here.
+			// Downgrade lock by taking read lock before returning to release
+			// write lock.
+			trace("Downgrading", writeLock());
+			readLock().lock();
```

Comment:
I see this is the recommended way but is it guaranteed not to block in all lock modes? Looks a bit scary. I think it would be nicer to use `tryLock()` to show (and guarantee) that we do not block here. An assert could be done on the return value to be true.

Reply:
If we cannot trust the documented way to downgrade a write lock, we are doomed. What would such an assert message say? "Java lied to us" :-)